[#11169] Student viewing responses: show an empty panel if the student didn't receive responses for a question

lyeyixian opened the pull request:
Fixes #11169

![image](https://user-images.githubusercontent.com/58477697/121668136-94534380-cadd-11eb-9137-0d37e0dcf800.png)

Adds a panel to tell the users that there are either no responses or the responses are not visible.

@damithc But, I'm not sure if I understand the issue correctly. I don't know if you want to have this message for each questions (multiple panels) or for the session as a whole (just one panel)?

damithc commented:
> @damithc But, I'm not sure if I understand the issue correctly. I don't know if you want to have this message for each questions (multiple panels) or for the session as a whole (just one panel)?

I think best to have it for each question. It's more informative that way.

lyeyixian commented:
>
> I think best to have it for each question. It's more informative that way.

OKOK, I will work on it.

moziliar commented:
@lyeyixian do you wanna mark your PR as WIP and remove WIP when it's ready for review? Thanks:)

lyeyixian commented:
> @lyeyixian do you wanna mark your PR as WIP and remove WIP when it's ready for review? Thanks:)

Yup, yes please. But, I don't know how to mark it as WIP, can you help me to do it instead, thank you.

damithc commented:
> Yup, yes please. But, I don't know how to mark it as WIP, can you help me to do it instead, thank you.

When the WIP feature was introduced by GitHub, this choice was only available at the point of creating a new PR (i.e., cannot got Normal -> WIP direction later). Not sure if GitHub allows it now.

damithc commented:
> When the WIP feature was introduced by GitHub, this choice was only available at the point of creating a new PR (i.e., cannot got Normal -> WIP direction later). Not sure if GitHub allows it now.

Apparently, it is possible to change a PR to a 'draft' now. 
![image](https://user-images.githubusercontent.com/1673303/121894043-3fb7fe80-cd51-11eb-860d-4ea3b8b56eb7.png)

lyeyixian commented:
Oh okok, then I will just change it to draft first in the mean time.

lyeyixian commented:
I faced some difficulties implementing the new feature that prof suggested, I need some help. 

![image](https://user-images.githubusercontent.com/58477697/122073735-3a2ee700-ce2b-11eb-8d33-7fe7ce4a354f.png)

We get the list of questions from the `getFeedbackSessionResults()` function, but it only includes questions that are visible and I can't seem to find an easy way to get the complete list of questions. I think it might have something to do with the backend side, so that when I called `getFeedbackSessionResults()` it returns the full complete list instead of only the visible ones. But, I can't find the exact file that do this and am currently stuck. 

I would like to ask if anyone can give me some pointer like which class or file should I look into on the backend side. Or is there other better way of doing this?

I'm also thinking if the above issue should be addressed in a separate PR instead? Might need your input on this @damithc .

damithc commented:
> I'm also thinking if the above issue should be addressed in a separate PR instead? Might need your input on this @damithc .

Let's wait to see if any senior dev can shed some light.

moziliar commented:
Hi @lyeyixian, first I would teach you how to navigate to the backend:)

When you see such a frontend request, you can use IDE to navigate to the function implementation (else just look up the reference and go to the file for the exact matched function via search); there, you would see the actual HTTP request, usually exposed to you as the backend endpoint form. In Intellij, you could just perform global keyword search (in this case I did `"/result"` as "/webapi"` is prepended with const) and go to the backend REST API map. Look for the router path -> look for the exact class it maps to -> go to the class. Now you are at `GetSessionResultsAction` class.

In response to your question about the logic: it is indeed intended that we only show all feedback questions for instructors and only the visible ones for the students.

Let me know if you need any further help! Thanks for contributing.

damithc commented:
> In response to your question about the logic: it is indeed intended that we only show all feedback questions for instructors and only the visible ones for the students.

There are two cases here:
1. The responses to the question is supposed to be visible to students, but the student has not received responses.
2. The responses for the question is not visible to students, but students see that question during submission
3. The question itself is not visible to students as they are not the intended respondents.

I think it is more informative to show 1 and 2 (but not 3) to the students in the results (more so for 1), even if there are no responses to show. a. We are not revealing anything they are not supposed to know. b. It is easier for the students to match results with the submission form they saw earlier if the questions are the same c. It is better if the student can distinguish between 'no responses' vs 'not visible'.

BTW, aren't we showing student's own responses for 1 and 2 anyway?

An (easier) alternative is to detect where there is nothing to show at all (like #11169) and insert an explanatory message to say there is nothing to show because either no responses or not visible. Optionally, we can also include a link to view the submission form so that student can easily go check the questions in the submission form -- especially if this doesn't require additional API calls.

lyeyixian commented:
> In response to your question about the logic: it is indeed intended that we only show all feedback questions for instructors and only the visible ones for the students.

Thank you. Hmmm okok, I see if I can work around without touching the backend.

lyeyixian commented:
OK, here's what I have done. @damithc can you take a look if this is ok?

**Original** 
Response not visible:
![image](https://user-images.githubusercontent.com/58477697/122186050-3600d900-cec0-11eb-84ce-d089725381dd.png)

**Updated**
Response not visible: 
![image](https://user-images.githubusercontent.com/58477697/122185575-c8ed4380-cebf-11eb-94fd-b799e4747a25.png)

Response visible and received: 
![image](https://user-images.githubusercontent.com/58477697/122185655-d9052300-cebf-11eb-83e6-519c9b71cefe.png)

Response visible but not received yet:
![image](https://user-images.githubusercontent.com/58477697/122186389-94c65280-cec0-11eb-9744-73175887756b.png)

I also make some changes to my previous implementation. 
![image](https://user-images.githubusercontent.com/58477697/122188640-bcb6b580-cec2-11eb-91b5-88a010654053.png)

damithc commented:
@lyeyixian 
* What happens if the student hasn't given responses? Note that the original issue was about the student seeing a completely blank space where session results should be i.e., no panels whatsoever
* Are you able to distinguish between 'not visible' vs 'visible but no responses'. If yes, we can make the text more specific to the exact reason.
* Probably no need for a panel. Can just show text only (?)
  ![image](https://user-images.githubusercontent.com/1673303/122208453-e3cab280-ced5-11eb-8029-5368c7a53bf2.png)

lyeyixian commented:
> * What happens if the student hasn't given responses? Note that the original issue was about the student seeing a completely blank space where session results should be i.e., no panels whatsoever

Currently, the question itself will not be shown, it will appear only after you have given response. 

If the student haven't submit his own response, the question data will not be added into the array, that's why it is not shown on the website. I still can't figure out a way to get the question as the API will only retrieve relevant question (if the student give his own response already, then it will show, but if he hasn't, then it will not be shown), the backend is written to work that way.

> * Are you able to distinguish between 'not visible' vs 'visible but no responses'. If yes, we can make the text more specific to the exact reason.
> * Probably no need for a panel. Can just show text only (?)

OKOK, I will change it.

damithc commented:
> If the student haven't submit his own response, the question data will not be added into the array, that's why it is not shown on the website. I still can't figure out a way to get the question as the API will only retrieve relevant question (if the student give his own response already, then it will show, but if he hasn't, then it will not be shown), the backend is written to work that way.

I see. It's something we have to solve. Might be able to reuse an API that is meant for generating the submission page (that page needs to show all the questions) but it will increase the API calls and database reads. Alternatively, we might have to modify the current backend API endpoint to send back the missing info as well. 

An easier fix that addresses the specific situation in the original issue : Detect cases where there is nothing to show and instead of a blank page, show a message informing the user that no responses have been given/received, or the received responses are not visible. I'm OK with that solution too. It's cheaper in terms of development cost as well as operations cost.

lyeyixian commented:
Is this ok? @damithc 

Not visible:
![image](https://user-images.githubusercontent.com/58477697/122363126-48941480-cf8b-11eb-9e7e-1873e798e16e.png)

Visible but no response:
![image](https://user-images.githubusercontent.com/58477697/122363155-4d58c880-cf8b-11eb-86e0-6615ebde3f7d.png)

lyeyixian commented:
> I see. It's something we have to solve. Might be able to reuse an API that is meant for generating the submission page (that page needs to show all the questions) but it will increase the API calls and database reads. Alternatively, we might have to modify the current backend API endpoint to send back the missing info as well.
> 
> An easier fix that addresses the specific situation in the original issue : Detect cases where there is nothing to show and instead of a blank page, show a message informing the user that no responses have been given/received, or the received responses are not visible. I'm OK with that solution too. It's cheaper in terms of development cost as well as operations cost.

Oh okok, then I put the PR back to ready for review?

damithc commented:
> Is this ok? @damithc
> 
> Not visible:
> ![image](https://user-images.githubusercontent.com/58477697/122363126-48941480-cf8b-11eb-9e7e-1873e798e16e.png)
> 
> Visible but no response:
> ![image](https://user-images.githubusercontent.com/58477697/122363155-4d58c880-cf8b-11eb-86e0-6615ebde3f7d.png)

Yes, looks fine.

damithc commented:
> Oh okok, then I put the PR back to ready for review?

Yes, after you have implemented the easier fix (unless you are going for the harder fix). Otherwise this PR doesn't really fix #11169 I've updated the issue description to clarify it further.

The fix shown in the recent screenshots are good to have but doesn't address the more severe case mentioned in the issue.

lyeyixian commented:
> Yes, after you have implemented the easier fix (unless you are going for the harder fix). Otherwise this PR doesn't really fix #11169 I've updated the issue description to clarify it further.

Oh, the easier fix actually is already been implemented.

I will update the text to the one you specified in https://github.com/TEAMMATES/teammates/issues/11169

![image](https://user-images.githubusercontent.com/58477697/122365976-bb9e8a80-cf8d-11eb-982c-7f9997d0fb14.png)

lyeyixian commented:
> The fix shown in the recent screenshots are good to have but doesn't address the more severe case mentioned in the issue.

So, do I leave it there or should I remove the updated code if this doesn't fix the issue?

damithc commented:
> Oh, the easier fix actually is already been implemented.
> 
> I will update the text to the one you specified in #11169
> 
> ![image](https://user-images.githubusercontent.com/58477697/122365976-bb9e8a80-cf8d-11eb-982c-7f9997d0fb14.png)

Got it. This text is fine too. It's shorter and not less informative than the one I suggested.

lyeyixian commented:
OKOK, then I will leave the text as is.

damithc commented:
Perhaps for all other cases we can show the following message at the bottom of the page?

`Note: Questions without responses (i.e., no responses received or responses are not meant to be visible to you) are not shown above.`

That will explain any missing questions from the page.

damithc commented:
> > The fix shown in the recent screenshots are good to have but doesn't address the more severe case mentioned in the issue.
> 
> So, do I leave it there or should I remove the updated code if this doesn't fix the issue?

Can keep. The fix is still tangentially related to the original issue and adds value.

lyeyixian commented:
> Perhaps for all other cases we can show the following message at the bottom of the page?
> 
> `Note: Questions without responses (i.e., no responses received or responses are not meant to be visible to you) are not shown above.`
> 
> That will explain any missing questions from the page.

Do you think it's better to put it at the top? It is also more consistent with other pages.

Like this.
![image](https://user-images.githubusercontent.com/58477697/122371163-3cf81c00-cf92-11eb-8ca9-3e901ca75f53.png)

lyeyixian commented:
Here's the changes.

![image](https://user-images.githubusercontent.com/58477697/122374915-58185b00-cf95-11eb-971a-69e3e546b150.png)

damithc commented:
> Here's the changes.
> 
> ![image](https://user-images.githubusercontent.com/58477697/122374915-58185b00-cf95-11eb-971a-69e3e546b150.png)

Looks fine @lyeyixian

lyeyixian commented:
Hi, this PR is already done and can proceed to be reviewed.

rrtheonlyone commented on src/web/app/pages-session/session-result-page/session-result-page.component.html:
```diff
@@ -43,6 +46,11 @@ <h1>
 <tm-loading-retry [shouldShowRetry]="hasFeedbackSessionResultsLoadingFailed" [message]="'Failed to load results'" (retryEvent)="retryLoadingFeedbackSessionResults()">
   <div *tmIsLoading="isFeedbackSessionResultsLoading">
+    <div *ngIf="isFeedbackSessionResultsEmpty" class="mt-4">
+      <div class="alert alert-primary d-flex align-items-center justify-content-center" style="background-color: #D1ECF1;" role="alert">
```
I am not sure if inline styles like`background-color` and `margin` are needed here.. @ccyccyccy can you help to verify?

lyeyixian replied:
The `background-color` might be unnecessary, but without the `margin` the text will be a bit off.
![image](https://user-images.githubusercontent.com/58477697/126892398-8a427d32-bdc6-4617-a635-222656825e5e.png)

lyeyixian replied:
OK, I figured out that the margin is not needed. I removed the `<p>` tag and it looks fine.

rrtheonlyone commented on src/web/app/pages-session/session-result-page/session-result-page.component.html:
```diff
@@ -55,6 +63,12 @@ <h1>
           <tm-student-view-responses *ngIf="responsesForOtherRecipient.length" [responses]="responsesForOtherRecipient" [feedbackQuestion]="question.feedbackQuestion" [timezone]="session.timeZone" [statistics]="question.questionStatistics"></tm-student-view-responses>
         </div>
         <div id="given-responses" class="mt-4" *ngIf="question.responsesFromSelf.length">
+          <div class="other-response" *ngIf="!question.feedbackQuestion.showResponsesTo.includes(visibilityRecipient)">
```
this is parked in the wrong place - you just need to do an else off this `<div id="other-responses" *ngIf="canStudentSeeResponses(question)">` on line 57. 

You also do not need the `question.feedbackQuestion.showResponsesTo.includes(visibilityRecipient)`, we can use `canStudentSeeResponses`

If not you will have issues like this where I set the responses to be actually anonymously visible to STUDENTS:

<img width="1077" alt="Screenshot 2021-07-25 at 1 10 33 PM" src="https://user-images.githubusercontent.com/10579415/126888486-eff0f769-f1d9-4b7a-97b9-ca1b88d20a2b.png">

lyeyixian replied:
Ah, didn't know we have this function `canStudentSeeResponses`. Thanks for the feedback, I will update it.

rrtheonlyone commented on src/web/app/pages-session/session-result-page/session-result-page.component.ts:
```diff
@@ -183,6 +185,9 @@ export class SessionResultPageComponent implements OnInit {
                 (a: QuestionOutput, b: QuestionOutput) =>
                     a.feedbackQuestion.questionNumber - b.feedbackQuestion.questionNumber);
+            if (this.questions.length === 0) {
+              this.isFeedbackSessionResultsEmpty = true;
```
why create this new variable? We can just check for `this.questions.length == 0` in the html right?

lyeyixian replied:
Yup, right. I didn't know about it before. I will change it.

lyeyixian replied:
OK, I just realised in #11271, the issue occurred because the `*ngIf` is a bit buggy when straight up using `this.someArray.length === 0`. I just tested it out, and the same problem occurs here. I will change back to using this variable first, if you have any other way to solve this can let me know.

xpdavid replied:
You can change the condition to `!isFeedbackSessionResultsLoading && this.someArray.length === 0` to avoid the flushing issue.

damithc replied:
On a separate note, glad to see you back in action again @xpdavid Welcome back!

rrtheonlyone replied:
Thanks @xpdavid ! The change above fixes this issue

rrtheonlyone reviewed (changes requested):
@lyeyixian thanks for the work here - let's also update `session-result-page.component.spec.ts` with some tests for the new behaviour you have added here. 

Some comments below:

lyeyixian commented:
> @lyeyixian thanks for the work here - let's also update `session-result-page.component.spec.ts` with some tests for the new behaviour you have added here.

I have added test for the new behaviour. But, I'm not sure if I'm doing it the right way and would like to have some help here.

rrtheonlyone reviewed (changes requested):
@lyeyixian there is progress - before moving forward, let's try to reach a consensus on the logic. 

#### Case 1: API returns an empty response 

We show this message "Nothing to show. The responses are either not configured to be visible or there are no responses.". Good that is correct. The 2 scenarios in which this can happen are mentioned.

#### Case 2: API returns questions

##### Case 2a: No responses are shown only a blank question. 

This happens when `questionDetails.isIndividualResponsesShownToStudents()` is false as in the case of a `FeedbackContributionQuestionDetails` question.

https://github.com/TEAMMATES/teammates/blob/f16e08e72e2e4040ebf16c85ebda8d56f79e7259/src/main/java/teammates/ui/output/SessionResultsData.java#L87

I believe we need to print a message saying "Individual Responses are not configured to be shown for this question type.

##### Case 2b: We only see responses from self. Responses to self is empty. 

In this case we still need to show Responses to Self (in your PR, it is hidden). We then need to put a text under "Other Responses (to you): Responses are not visible to you". You have this in place.

This happens when `questionDetails.isIndividualResponsesShownToStudents()` is true but the visibility settings are such that only your own responses are shown. You can verify this in the backend as an exercise.

### Miscellaneous UI 
- I like the note at the top "Note: Questions without responses (i.e., no responses received or responses are not meant to be visible to you) are not shown below." However, it is an easy change to let the backend pass questions too so we can reduce the number of cases here. As mentioned above, @damithc we can create a separate issue for this to not complicate this PR.

@lyeyixian if you agree with the above, let's start on a clean slate and do the above logic. Right now the if statements look a little chaotic to me and I already spotted some errors when testing. Once you have the logic in place, you can proceed to add the snapshot tests. (the testing plan you laid out looks good). Also remember to make sure CI passes. 

@damithc for your review on the above plan too

damithc commented:
> @damithc for your review on the above plan too

Sounds good. You can go ahead @lyeyixian 
Thanks for looking into this @rrtheonlyone and thanks for the work put into this PR so far @lyeyixian

lyeyixian commented:
> #### Case 2a: No responses are shown only a blank question.
> This happens when `questionDetails.isIndividualResponsesShownToStudents()` is false as in the case of a `FeedbackContributionQuestionDetails` question.
> 
> https://github.com/TEAMMATES/teammates/blob/f16e08e72e2e4040ebf16c85ebda8d56f79e7259/src/main/java/teammates/ui/output/SessionResultsData.java#L87
> 
> I believe we need to print a message saying "Individual Responses are not configured to be shown for this question type.
> 

Do you mean to add the message somewhere in the question panel?

![image](https://user-images.githubusercontent.com/58477697/127989412-7aa519b7-64b3-4165-a06c-ebb769f18ff7.png)

lyeyixian commented:
@rrtheonlyone Sorry, I'm still quite confused as to what I need to do here.
From what I see, I have already done case 1 and case 2b. For case 2a, I should add a message (the question I asked above)

>@lyeyixian if you agree with the above, let's start on a clean slate and do the above logic. Right now the if statements look a little chaotic to me and I already spotted some errors when testing. Once you have the logic in place, you can proceed to add the snapshot tests. (the testing plan you laid out looks good). Also remember to make sure CI passes.

Or is it the if statement is wrong or too messy? So, I need to rethink the logic and follow the 3 cases that you have mentioned?

>**Case 2b: We only see responses from self. Responses to self is empty.**
>
>In this case we still need to show Responses to Self (in your PR, it is hidden). We then need to put a text under "Other Responses (to you): Responses are not visible to you". You have this in place.

I don't quite get this. If there is no responses or the responses shouldn't be seen, why would we still need to show Responses to Self. Or you mean we still need to have the response panel, just that it is empty? Then put the current text that I have under it.

rrtheonlyone commented:
Hi @lyeyixian, 

Let's not worry too much about implementation first. It is good you already have done parts of it. Let's get on the same page for the below.

> Do you mean to add the message somewhere in the question panel?

Yes, maybe show it below the "Comparison of Work Distribution"

> Or is it the if statement is wrong or too messy? So, I need to rethink the logic and follow the 3 cases that you have mentioned?

It looks a little messy at the moment - we could try to clean it up based on the above proposal.

> I don't quite get this. If there is no responses or the responses shouldn't be seen, why would we still need to show Responses to Self. Or you mean we still need to have the response panel, just that it is empty? Then put the current text that I have under it.

Ah typo here, I meant **Response From Self**. (updated the above description accordingly).

lyeyixian commented:
> It looks a little messy at the moment - we could try to clean it up based on the above proposal.

OK, I realised where the problem was and the cause of my confusion. I didn't check and look through carefully when making changes based on this review https://github.com/TEAMMATES/teammates/pull/11190#discussion_r676085224 and assumed everything will look the same since the logic didn't change much, I'm just refactoring the code. 

Now, I realised my mistake and already updated it.

lyeyixian commented:
> Yes, maybe show it below the "Comparison of Work Distribution"

Do you think it is ok to put it in the additional info? 

![image](https://user-images.githubusercontent.com/58477697/128749130-e7b15a7f-f4d9-43cd-b583-ea4c44c69489.png)

![image](https://user-images.githubusercontent.com/58477697/128749018-0d4c76bb-4633-4d5f-8eee-18f575417df4.png)

damithc commented:
> > Yes, maybe show it below the "Comparison of Work Distribution"
> 
> Do you think it is ok to put it in the additional info?

Given that this is added by us (in contrast, additional info of the question is added by the instructor), it is better the two appears distinct from each other. Perhaps separate from the rest with a blank line? Can also add in info icon in front.

lyeyixian commented:
> Given that this is added by us (in contrast, additional info of the question is added by the instructor), it is better the two appears distinct from each other. Perhaps separate from the rest with a blank line? Can also add in info icon in front.

OKOK

@damithc Do you mean below the box? Like this?

![image](https://user-images.githubusercontent.com/58477697/129049089-06c9fbc1-9e28-4e26-9e13-219c42258322.png)

damithc commented:
> @damithc Do you mean below the box? Like this?
> 
> ![image](https://user-images.githubusercontent.com/58477697/129049089-06c9fbc1-9e28-4e26-9e13-219c42258322.png)

Yes, that's fine too @lyeyixian

daongochieu2810 commented:
@lyeyixian have you tried this on staging env? And with Firefox?

lyeyixian commented:
Nope, I haven't. Yea, I'm using Firefox to do the test.

lyeyixian commented:
@daongochieu2810 when I run the e2e test on the staging server, it failed just like what the github action showed

daongochieu2810 commented:
@lyeyixian I pulled your repo and tested locally. Can you try again?
![Screenshot from 2021-08-14 14-51-01](https://user-images.githubusercontent.com/53283766/129437900-f35b3d20-cbfe-4482-ace8-285fb9e93fea.png)

lyeyixian commented:
@daongochieu2810 I found the solution. 

Apparently, I missed out something causing the number scale statistic to be hidden. 

I think the reason why when I run the test locally it works fine is because e2e test runs on frontend files that are built not served? So, it still run the test on old code not the updated one. That's why when I do the staging as you suggested, the test result and the code is in sync again. 

It should pass the test now.

rrtheonlyone commented on src/web/app/pages-session/session-result-page/session-result-page.component.html:
```diff
@@ -43,21 +46,39 @@ <h1>
+                              [responses]="question.allResponses"
+                              [statistics]="question.questionStatistics"
+                              [isStudent]="true"></tm-single-statistics>
+          <div *ngIf="question.questionStatistics, else dontShowAdditionalMsg">
```
This `question.questionStatistics` check looks a little hardcoded and I am not sure if this is even correct. We should expose this field somehow in the API return:
```
    public boolean isIndividualResponsesShownToStudents() {
        return true;
```

However, I am not sure if this addition is easy. We can use a constant map otherwise too for now and KIV this.

lyeyixian replied:
I'm also not sure how I can go about doing this. May I ask what is a constant map that you said? I can maybe try to implement that.

rrtheonlyone replied:
just have a class variable in `session-result-page-component.ts` with this comment:
```
//isIndividualResponsesShownToStudents() is false for these questions
RESPONSE_HIDDEN_QUESTIONS = [
      "CONTRIB"
]
```

Then check if question.feedbackQuestion.questionType is in the array above

rrtheonlyone reviewed (changes requested):
Almost there, with reference to the above changes requested:

## Case 1: API Returns an empty response

<img width="1154" alt="Screenshot 2021-08-22 at 3 29 11 AM" src="https://user-images.githubusercontent.com/10579415/130333142-f3a8a17f-4b3f-4285-aedb-d555e8f2b586.png">

Let's remove the toolbar at the top for this case @lyeyixian 

## Case 2: API returns questions 

### Case 2a: No responses are shown only a blank question 

<img width="1096" alt="Screenshot 2021-08-22 at 3 26 57 AM" src="https://user-images.githubusercontent.com/10579415/130333151-660d0fa3-fa92-4c45-98b5-a7482213fe55.png">


### Case 2b: We only see responses from self. Responses to self is empty.

<img width="1078" alt="Screenshot 2021-08-22 at 3 27 38 AM" src="https://user-images.githubusercontent.com/10579415/130333155-ab2150c6-41c8-4103-aeb6-3f41125f492e.png">

### Misc: Toolbar at the top (always present)

<img width="1110" alt="Screenshot 2021-08-22 at 3 28 22 AM" src="https://user-images.githubusercontent.com/10579415/130333158-3f5e062c-dc65-4ca0-a291-c28bae4fab91.png">


Tagging @damithc in case there are further comments on UI here

lyeyixian commented:
> ## Case 1: API Returns an empty response
> <img alt="Screenshot 2021-08-22 at 3 29 11 AM" width="1154" src="https://user-images.githubusercontent.com/10579415/130333142-f3a8a17f-4b3f-4285-aedb-d555e8f2b586.png">
> 
> Let's remove the toolbar at the top for this case @lyeyixian
> 

OK, removed it.

rrtheonlyone commented:
@lyeyixian you may have committed unnecessary files into this PR

lyeyixian commented:
@rrtheonlyone I have added the thing you said. Can you help me see if it is the correct way of doing it?



> @lyeyixian you may have committed unnecessary files into this PR

Sorry, I reverted the changes. I updated it last time because the editor was throwing some error and I thought I have to change them.

lyeyixian commented:
@rrtheonlyone Hi, can you review my PR?

rrtheonlyone commented:
@lyeyixian thanks for making the changes and apologies for the long delay on this! I am looking through your changes and will review by end of the week latest. Just want to run locally and ensure the different cases are okay: I have no major issues with your code and it looks good!

rrtheonlyone reviewed (approved):
LGTM 💯 Great Work @lyeyixian

I did a very minor change here: https://github.com/TEAMMATES/teammates/pull/11190/commits/05f4c7f13f3a427dae39246cbe087d74e5efbb64 that removes `isFeedbackSessionResultsEmpty`
